Add proposal for device management enhancement (#1687)
kubeedge/kubeedge: docs/proposals/device-management-enhance.md

    model: cc2650-sensortag
spec:
  deviceModelRef:
    name: cc2650-sensortag
  protocol:
    comom:
      com:
        serialPort: '1'
        baudRate: 115200
        dataBits: 8
        parity: even
        stopBits: 1
      commType: 0
    customizedProtocol:
      protocolName: MY-TEST-PROTOCOL
      configData:
        key1: val1
        key2: val2
        key3:
          innerKey1: ival1
  nodeSelector:
    nodeSelectorTerms:
    - matchExpressions:
      - key: ''
        operator: In
        values:
        - edge-node1          #pls give your edge node name
  propertyVisitors:
    - propertyName: temperature
      collectCycle: 500000000
      reportCycle: 1000000000
      customizedProtocol:
        protocolName: MY-TEST-PROTOCOL
        configData:
          def1: def1-val
          def2: def2-val
          def3:
            innerDef1: idef-val
    - propertyName: temperature-enable
      collectCycle: 500000000
      reportCycle: 1000000000
      bluetooth:
        characteristicUUID: f000aa0204514000b000000000000000
        dataWrite:
        "ON": [1]
        "OFF": [0]
status:
  twins:
    - propertyName: temperature-enable
    - propertyName: io-data

> data:

Reviewer: data should be under `spec`

Author: done

This proposal addresses how can we enhance device management including more flexible
protocol support, more reasonable device model and device instance design, and
additional way to handle data which is collected from device.

### Goals

* Add customized protocol support
* Add ability to allow user get and process data in edge node
* Improve device model and device instance crd design

### Non-goals

* To support streaming data from video device
* To provide specific protocols and mappers

## Proposal

We propose below modifications on current design.
* Move property visitors from device model to device instance.
* Add collectCycle and reportCycle under property visitor.
* Add data section in twin spec.
* Add customized protocol config in device model and device instance CRDs.
* Add common part in protocol config section
* Extract common part of Modbus protocol config into an independent common part.
* Allow add any customized K-V in protocol config and property visitor section.

### Use Cases

* Reuse device model.
  * Considering device properties are physical attributes, but property visitors are manually configured

> attributes. Combining device properties and property visitors in device model reduce the reusability of device model.

Reviewer: "Combine device properties and property visitors" or "remove property visitors to device instance"?

Author: It is an example for existing device model design. 

"Combining device properties and property visitors in device model _**decrease**_ the reusability of device model."

>   * Case 1: Same devices are connected to a central management server, eg. SCADA. In this case, devices have same properties but

Reviewer: Just a small question :)  Same devices are connected to a central management server, they will also have different property visitors?

  different property visitors.
* Customized data collect cycle and report cycle
   * Users can define collect cycle and report cycle to each property. For example, a temperature property may need be collected
   per second, while a throughput property may need be collected per hour.
* Deal data of non-twin properties.
  * Currently, only twin properties will be sync between edge and cloud. Non-twin properties are not processed by edge-core. 
  Time-Serial data are produced from devices and should have a way to allow user deal with these data.
* Deal various industrial protocols
  * Currently, only Modbus, OPC-UA and bluetooth are supported by KubeEdge. However there are thousands of industrial protocols.
  It is impossible to define all these protocols in KubeEdge. If users want to use these un-predefined protocols, we should provide
  a way to support.
* Customized provided protocol
  * If user want to add some special control value, such as bulk related collection, in provided protocol like Modbus, he can use the customized K-V features.

## Design Details

### Move property visitors from device model to device instance.

> - move property visitors of device CRD to device instance CRD

Reviewer: ```suggestion
- move property visitors from device model CRD to device instance CRD
```

> - move property visitors of DeviceModelSpec to DeviceSpec struct.

Reviewer: from

	// Required.
	Port int64 `json:"port,omitempty"`
}

type ProtocolConfigCOM struct {
	...
	// Required.
	SerialPort string `json:"serialPort,omitempty"`
	// Required. BaudRate 115200|57600|38400|19200|9600|4800|2400|1800|1200|600|300|200|150|134|110|75|50
	Parity string `json:"parity,omitempty"`
	// Required. Bit that stops 1|2
	StopBits int64 `json:"stopBits,omitempty"`
}

type ProtocolConfigCustomized struct {
	// Unique protocol name
	// Required.
	ProtocolName string `json:"protocolName,omitempty"`
	// Any config data
	// +optional
	ConfigData *CustomizedValue `json:"configData,omitempty"`
}

// DeviceData reports the device's time-series data to edge MQTT broker.
// These data should not be processed by edgecore. Instead, they can be process by
// third-party data-processing apps like EMQX kuiper.
type DeviceData struct {
	// Required: A list of data properties, which are not required to be processed by edgecore
	DataProperties []DataProperty `json:"dataProperties,omitempty"`
	// Topic used by mapper, all data collected from dataProperties
	// should be published to this topic,
	// the default value is $ke/events/device/+/data/update
	// +optional

> 	DataTopic string `json:"dataTopic,omitempty"`

Reviewer: how does third-party data-processing apps like EMQX kuiper get the topic?

Author: third-party can get this info by using configMap generated by KubeEdge, or they can use Device API query this info and add it to their own configMap

	// Additional metadata like timestamp when the value was reported etc.
	// +optional
	Metadata map[string]string `json:"metadata,omitempty"`
}

// DevicePropertyVisitor describes the specifics of accessing a particular device
// property. Visitors are intended to be consumed by device mappers which connect to devices
// and collect data / perform actions on the device.
type DevicePropertyVisitor struct {
	// Required: The device property name to be accessed. This should refer to one of the
	// device properties defined in the device model.
	PropertyName string `json:"propertyName,omitempty"`
	// Define how frequent mapper will report the value.
	// +optional
	ReportCycle int64 `json:"reportCycle,omitempty"`
	// Define how frequent mapper will collect from device.
	// +optional
	CollectCycle int64 `json:"collectCycle,omitempty"`
	// Customized values for visitor of provided protocols
	// +optional
	CustomizedValues *CustomizedValue `json:"customizedValues,omitempty"`
	// Required: Protocol relevant config details about the how to access the device property.
	VisitorConfig `json:",inline"`
}
```

### modifications on device model types
```golang

> type VisitorConfig struct {

Reviewer: Does the VisitorConfig should be move to device instance? Since no one will use it in device model

---
title: Device Management Enhancement
authors:
    - "@luogangyi"

> approvers:

Reviewer 2: ```suggestion
approvers:
    - "@kevin-wangzefeng"
    - "@fisherxu"
```

Author: done

creation-date: 2020-05-13
last-updated:

> status:

Reviewer 2: ```suggestion
status: implementable
```

### Add collectCycle and reportCycle under property visitor.
- add collectCycle and reportCycle under property visitor in device instance CRD.
- add collectCycle and reportCycle in DevicePropertyVisitor struct.

### Add data section besides twin section.
- add data section in device instance CRD.
- add DeviceData in Device struct.
- inject data section and twin section into configmap
- add new MQTT topic to handle data from data section

### Add customized protocols support
- add 'other' under property visitor, type is object

### modifications on device instance types
```golang
type DeviceSpec struct {
	DeviceModelRef *v1.LocalObjectReference `json:"deviceModelRef,omitempty"`
	// Required: The protocol configuration used to connect to the device.
	Protocol ProtocolConfig `json:"protocol,omitempty"`
	// List of property visitors which describe how to access the device properties.
	// PropertyVisitors must unique by propertyVisitor.propertyName.
	// +optional
	PropertyVisitors []DevicePropertyVisitor `json:"propertyVisitors,omitempty"`
	// Data section describe a list of time-series properties which should be processed
	// on edge node.
	// +optional

>   Data DeviceData `json:"data,omitempty"`

Author: done

  ...
}

type ProtocolConfig struct {
  ...
	// Protocol configuration for bluetooth
	// +optional
	Bluetooth *ProtocolConfigBluetooth `json:"bluetooth,omitempty"`
	// Configuration for customized protocol
	// +optional

>   CustomizedProtocol *ProtocolConfigCustomized `json:"customizedProtocol,omitempty"`

Author: done

  ...
}

type ProtocolConfigCustomized struct {
	// Unique protocol name
	// Required.
  ProtocolName string           `json:"protocolName,omitempty"`
  // Any config data

>   // +optional

Author: done

	ConfigData   *CustomizedValue `json:"configData,omitempty"`
}

// DeviceData reports the device's time-series data to edge MQTT broker.
// These data should not be processed by edgecore. Instead, they can be process by
// third-party data-processing apps like EMQX kuiper.
type DeviceData struct {
	// A list of data properties, which are not required to be processed by edgecore
	// Required.
	DataProperties []DataProperty `json:"dataProperties,omitempty"`
	// Topic used by mapper, all data collected from dataProperties
	// should be published to this topic,
  // the default value is $ke/events/device/+/data/update

>   // +optional

Author: done

	DataTopic string `json:"dataTopic,omitempty"`
}

// DataProperty represents the device property for external use.
type DataProperty struct {
	// Required: The property name for which should be processed by external apps.
	// This property should be present in the device model.
	PropertyName string `json:"propertyName,omitempty"`
	// Additional metadata like timestamp when the value was reported etc.
	// +optional
	Metadata map[string]string `json:"metadata,omitempty"`
}

// DevicePropertyVisitor describes the specifics of accessing a particular device
// property. Visitors are intended to be consumed by device mappers which connect to devices
// and collect data / perform actions on the device.
type DevicePropertyVisitor struct {
	// Required: The device property name to be accessed. This should refer to one of the
	// device properties defined in the device model.
	PropertyName string `json:"propertyName,omitempty"`

>   // Define how frequent mapper will report the value. Nano Second.
>   // +optional

Reviewer 2: ```suggestion
	// Define how frequent mapper will report the value. Nano Second.
	// +optional
```

Author: done

	ReportCycle int64 `json:"reportCycle,omitempty"`

>   // Define how frequent mapper will collect from device. Nano Second.
>   // +optional

Reviewer 2: ```suggestion
	// Define how frequent mapper will collect from device. Nano Second.
	// +optional
```

Author: done

	CollectCycle int64 `json:"collectCycle,omitempty"`
	// Required: Protocol relevant config details about the how to access the device property.
	VisitorConfig `json:",inline"`
}
```

### modifications on device model types
```golang
type VisitorConfig struct {
  ...
	// Bluetooth represents a set of additional visitor config fields of bluetooth protocol.
	// +optional
	Bluetooth *VisitorConfigBluetooth `json:"bluetooth,omitempty"`
	// CustomizedProtocol represents a set of visitor config fields of bluetooth protocol.
	// +optional

>   CustomizedProtocol *VisitorConfigCustomized `json:"customizedProtocol,omitempty"`

Reviewer 2: ```suggestion
	CustomizedProtocol *VisitorConfigCustomized `json:"customizedProtocol,omitempty"`
```
